File: Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class cmp{
public:
    bool operator() (const ListNode *l, const ListNode *r) const {
        return l->val > r->val;
    }    
};
class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<ListNode *>::iterator iter = lists.begin();
        while (iter != lists.end()) {
            if (*iter == NULL) {
                lists.erase(iter);
            } else {
                iter++;
            }
        }
        if (lists.size() < 1) return NULL;
        
        ListNode *head = NULL, *cur = NULL;
        make_heap(lists.begin(), lists.end(), cmp());
        
        while (lists.size() > 0) {
            if (head == NULL) {
                head = cur = lists[0];
            } else {
                cur = cur->next = lists[0];
            }
            
            pop_heap(lists.begin(), lists.end(), cmp());
            int last = lists.size() - 1;
            if (lists[last]->next != NULL) {
                lists[last] = lists[last]->next;
                push_heap(lists.begin(), lists.end(), cmp());
            } else {
                lists.pop_back();
            }
        }
        return head;
    }
};

File: Merge k Sorted Lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists==null||lists.length==0) {
            return null;
        }
    
        PriorityQueue<ListNode> heap = new PriorityQueue<ListNode>(new Comparator<ListNode>(){
            public int compare(ListNode l1, ListNode l2){
                return l1.val - l2.val;
            }
        });
        for (ListNode list : lists) {
            if (list != null) {
                heap.offer(list);
            }
        }

        ListNode head = new ListNode(-1);
        ListNode p = head;
        while (!heap.isEmpty()) {
            ListNode cur = heap.poll();
            if (cur.next != null) {
                heap.offer(cur.next);
                cur.next = null;
            }
            p.next = cur;
            p = p.next;
        }
        
        return head.next;
    }
}
